In [2]:
import numpy as np

In [3]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris = np.genfromtxt(url, delimiter=',', dtype='str')
iris

array([['5.1', '3.5', '1.4', '0.2', 'Iris-setosa'],
       ['4.9', '3.0', '1.4', '0.2', 'Iris-setosa'],
       ['4.7', '3.2', '1.3', '0.2', 'Iris-setosa'],
       ['4.6', '3.1', '1.5', '0.2', 'Iris-setosa'],
       ['5.0', '3.6', '1.4', '0.2', 'Iris-setosa'],
       ['5.4', '3.9', '1.7', '0.4', 'Iris-setosa'],
       ['4.6', '3.4', '1.4', '0.3', 'Iris-setosa'],
       ['5.0', '3.4', '1.5', '0.2', 'Iris-setosa'],
       ['4.4', '2.9', '1.4', '0.2', 'Iris-setosa'],
       ['4.9', '3.1', '1.5', '0.1', 'Iris-setosa'],
       ['5.4', '3.7', '1.5', '0.2', 'Iris-setosa'],
       ['4.8', '3.4', '1.6', '0.2', 'Iris-setosa'],
       ['4.8', '3.0', '1.4', '0.1', 'Iris-setosa'],
       ['4.3', '3.0', '1.1', '0.1', 'Iris-setosa'],
       ['5.8', '4.0', '1.2', '0.2', 'Iris-setosa'],
       ['5.7', '4.4', '1.5', '0.4', 'Iris-setosa'],
       ['5.4', '3.9', '1.3', '0.4', 'Iris-setosa'],
       ['5.1', '3.5', '1.4', '0.3', 'Iris-setosa'],
       ['5.7', '3.8', '1.7', '0.3', 'Iris-setosa'],
       ['5.1

## iris는 넘파이 배열 객체입니다. 아래의 문제들에 대해서 코드를 작성하시고 실행시키세요. (iris 객체가 문제가 진행됨에 따라 변경되도록 코드를 실행시키세요)

## 1. iris의 차원, 모양, 크기를 프린트하세요

In [4]:
iris.ndim, iris.shape, iris.size

(2, (150, 5), 750)

## 2. iris의 1번 axis에서 4번 값들 (Iris-setosa,...,iris-virginica)을 없애고, 수정된 iris를 프린트하세요. ((150,4) 모양의 배열) (불리안 객체 기반 색인 혹은 팬시 색인을 활용하세요.)

In [5]:
iris = iris[:, [0, 1, 2, 3]] # iris = iris[:, [True, True, True, True, False]]
iris

array([['5.1', '3.5', '1.4', '0.2'],
       ['4.9', '3.0', '1.4', '0.2'],
       ['4.7', '3.2', '1.3', '0.2'],
       ['4.6', '3.1', '1.5', '0.2'],
       ['5.0', '3.6', '1.4', '0.2'],
       ['5.4', '3.9', '1.7', '0.4'],
       ['4.6', '3.4', '1.4', '0.3'],
       ['5.0', '3.4', '1.5', '0.2'],
       ['4.4', '2.9', '1.4', '0.2'],
       ['4.9', '3.1', '1.5', '0.1'],
       ['5.4', '3.7', '1.5', '0.2'],
       ['4.8', '3.4', '1.6', '0.2'],
       ['4.8', '3.0', '1.4', '0.1'],
       ['4.3', '3.0', '1.1', '0.1'],
       ['5.8', '4.0', '1.2', '0.2'],
       ['5.7', '4.4', '1.5', '0.4'],
       ['5.4', '3.9', '1.3', '0.4'],
       ['5.1', '3.5', '1.4', '0.3'],
       ['5.7', '3.8', '1.7', '0.3'],
       ['5.1', '3.8', '1.5', '0.3'],
       ['5.4', '3.4', '1.7', '0.2'],
       ['5.1', '3.7', '1.5', '0.4'],
       ['4.6', '3.6', '1.0', '0.2'],
       ['5.1', '3.3', '1.7', '0.5'],
       ['4.8', '3.4', '1.9', '0.2'],
       ['5.0', '3.0', '1.6', '0.2'],
       ['5.0', '3.4', '1.6', '0.4'],
 

## 3. iris의 데이터 타입을 np.float64로 바꾸고, 바꾼 후의 iris의 데이터 타입을 프린트하세요.

In [6]:
iris = iris.astype(np.float64)
iris.dtype

dtype('float64')

## 4. 유니버설 함수를 활용해서 iris의 소수부를 모두 버리고, 수정된 iris를 프린트하세요.

In [7]:
iris = np.floor(iris)
iris

array([[5., 3., 1., 0.],
       [4., 3., 1., 0.],
       [4., 3., 1., 0.],
       [4., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [4., 3., 1., 0.],
       [5., 3., 1., 0.],
       [4., 2., 1., 0.],
       [4., 3., 1., 0.],
       [5., 3., 1., 0.],
       [4., 3., 1., 0.],
       [4., 3., 1., 0.],
       [4., 3., 1., 0.],
       [5., 4., 1., 0.],
       [5., 4., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [4., 3., 1., 0.],
       [5., 3., 1., 0.],
       [4., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [4., 3., 1., 0.],
       [4., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 4., 1., 0.],
       [5., 4., 1., 0.],
       [4., 3., 1., 0.],
       [5., 3., 1., 0.],
       [5., 3., 1., 0.],
       [4., 3., 1., 0.],
       [4., 3., 1., 0.],
       [5., 3., 1., 0.],


## 5. iris와 같은 크기 (150,4) 이면서 표준정규분포를 따르는 noise 배열을 만들고 프린트하세요. (넘파이 random 모듈의 난수 generator를 생성하여 활용하고, 그 seed는 1000번으로 하세요.)

In [8]:
rng = np.random.default_rng(seed=1000)
noise = rng.standard_normal((150, 4))
noise

array([[-3.21330206e-01, -4.85661478e-01,  1.68005813e+00,
         1.97052688e+00],
       [ 1.55183661e-01, -1.25776545e+00,  2.00218415e-01,
         4.20529922e-02],
       [-8.78016468e-01,  1.52940790e+00,  9.95537174e-01,
        -1.40923974e+00],
       [ 8.83434150e-01, -1.42880669e+00, -4.33472757e-01,
        -8.88188047e-01],
       [-7.42407887e-01,  1.02469979e-03, -4.73708529e-01,
        -3.16114619e-01],
       [ 1.31339378e+00, -3.00980716e-01,  4.22731051e-01,
        -1.53514984e+00],
       [ 1.92553344e-01,  6.97102848e-01,  1.23542270e-01,
        -5.88134091e-01],
       [ 1.05105802e+00,  1.45268066e+00,  1.50100505e-01,
         1.59953006e+00],
       [ 1.31740169e+00,  2.82221904e-01,  3.49968050e-01,
         9.93568522e-01],
       [ 6.17700418e-01, -1.04310039e+00, -1.81087023e-01,
         1.20386544e+00],
       [-1.39844198e+00, -1.12643770e-02, -7.26997128e-01,
         3.71771445e-01],
       [ 7.65718663e-01,  1.23352496e+00,  1.64677645e+00,
      

## 6. iris와 noise를 더한뒤, 1번 axis방향으로 누적합으로 iris를 수정하고, 프린트하세요. (크기가 (150,4)인 넘파이 배열)

In [9]:
iris = np.cumsum(iris+noise, axis=1)
iris

array([[ 4.67866979,  7.19300832,  9.87306644, 11.84359333],
       [ 4.15518366,  5.89741821,  7.09763663,  7.13968962],
       [ 3.12198353,  7.65139143,  9.6469286 ,  8.23768887],
       [ 4.88343415,  6.45462746,  7.0211547 ,  6.13296666],
       [ 4.25759211,  7.25861681,  7.78490828,  7.46879366],
       [ 6.31339378,  9.01241306, 10.43514411,  8.89999427],
       [ 4.19255334,  7.88965619,  9.01319846,  8.42506437],
       [ 6.05105802, 10.50373868, 11.65383918, 13.25336925],
       [ 5.31740169,  7.59962359,  8.94959164,  9.94316017],
       [ 4.61770042,  6.57460003,  7.39351301,  8.59737844],
       [ 3.60155802,  6.59029364,  6.86329651,  7.23506796],
       [ 4.76571866,  8.99924362, 11.64602007, 12.94686899],
       [ 3.48668664,  6.56532747,  8.18133914,  7.77050852],
       [ 3.57060154,  5.73676237,  7.48502935,  6.28495369],
       [ 5.8314477 ,  9.28205553, 12.58024457, 12.01351974],
       [ 5.23576504, 10.2080982 ,  9.79457819,  8.89509187],
       [ 5.11172638,  7.

## 7. np.where를 활용해서 iris의 값 중 10을 넘어가는 값은 10으로 수정하여 프린트하세요.

In [10]:
iris = np.where(iris>10, 10, iris) # 10이 넘어가면 0으로 수정, 안 넘어가면 기존 값 유지
iris

array([[ 4.67866979,  7.19300832,  9.87306644, 10.        ],
       [ 4.15518366,  5.89741821,  7.09763663,  7.13968962],
       [ 3.12198353,  7.65139143,  9.6469286 ,  8.23768887],
       [ 4.88343415,  6.45462746,  7.0211547 ,  6.13296666],
       [ 4.25759211,  7.25861681,  7.78490828,  7.46879366],
       [ 6.31339378,  9.01241306, 10.        ,  8.89999427],
       [ 4.19255334,  7.88965619,  9.01319846,  8.42506437],
       [ 6.05105802, 10.        , 10.        , 10.        ],
       [ 5.31740169,  7.59962359,  8.94959164,  9.94316017],
       [ 4.61770042,  6.57460003,  7.39351301,  8.59737844],
       [ 3.60155802,  6.59029364,  6.86329651,  7.23506796],
       [ 4.76571866,  8.99924362, 10.        , 10.        ],
       [ 3.48668664,  6.56532747,  8.18133914,  7.77050852],
       [ 3.57060154,  5.73676237,  7.48502935,  6.28495369],
       [ 5.8314477 ,  9.28205553, 10.        , 10.        ],
       [ 5.23576504, 10.        ,  9.79457819,  8.89509187],
       [ 5.11172638,  7.

In [11]:
import pandas as pd
from pandas import Series, DataFrame
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris = np.genfromtxt(url, delimiter=',', dtype='str')

## 8. column을 순서대로 'a','b','c','d','e' 로 갖는 데이터프레임의 형태로 iris를 만들고 프린트하세요.

In [12]:
iris = DataFrame(iris, columns=["a", "b", "c", "d", "e"])
iris

a    b    c    d               e
0    5.1  3.5  1.4  0.2     Iris-setosa
1    4.9  3.0  1.4  0.2     Iris-setosa
2    4.7  3.2  1.3  0.2     Iris-setosa
3    4.6  3.1  1.5  0.2     Iris-setosa
4    5.0  3.6  1.4  0.2     Iris-setosa
..   ...  ...  ...  ...             ...
145  6.7  3.0  5.2  2.3  Iris-virginica
146  6.3  2.5  5.0  1.9  Iris-virginica
147  6.5  3.0  5.2  2.0  Iris-virginica
148  6.2  3.4  5.4  2.3  Iris-virginica
149  5.9  3.0  5.1  1.8  Iris-virginica

[150 rows x 5 columns]

## 9. iris의 index에 "samples', columns에 'information'이라는 name 속성을 주고, iris를 프린트하세요.

In [14]:
iris.index.name="samples"
iris.columns.name="information"
iris

information    a    b    c    d               e
samples                                        
0            5.1  3.5  1.4  0.2     Iris-setosa
1            4.9  3.0  1.4  0.2     Iris-setosa
2            4.7  3.2  1.3  0.2     Iris-setosa
3            4.6  3.1  1.5  0.2     Iris-setosa
4            5.0  3.6  1.4  0.2     Iris-setosa
...          ...  ...  ...  ...             ...
145          6.7  3.0  5.2  2.3  Iris-virginica
146          6.3  2.5  5.0  1.9  Iris-virginica
147          6.5  3.0  5.2  2.0  Iris-virginica
148          6.2  3.4  5.4  2.3  Iris-virginica
149          5.9  3.0  5.1  1.8  Iris-virginica

[150 rows x 5 columns]

## 10. reindex를 활용하여, iris의 columns 중에서는 'a','c','e' 와 index는 짝수만 갖도록 수정하고, 프린트하세요.

In [15]:
iris = iris.reindex(columns=["a", "c", "e"], index=np.arange(0,150,2))
iris

information    a    c               e
samples                              
0            5.1  1.4     Iris-setosa
2            4.7  1.3     Iris-setosa
4            5.0  1.4     Iris-setosa
6            4.6  1.4     Iris-setosa
8            4.4  1.4     Iris-setosa
...          ...  ...             ...
140          6.7  5.6  Iris-virginica
142          5.8  5.1  Iris-virginica
144          6.7  5.7  Iris-virginica
146          6.3  5.0  Iris-virginica
148          6.2  5.4  Iris-virginica

[75 rows x 3 columns]

## 11. iris의 'a' 피처와 'c' 피처를 astype을 활용하여, np.float64로 형변환을 진행하고, loc을 활용하여, 'a'피처의 값이 5를 넘는 sample들만 남도록하세요, 수정된 iris를 프린트하세요.


In [17]:
iris[['a', 'c']] = iris[['a', 'c']].astype(np.float64)

In [21]:
iris = iris.loc[iris.a>5, :] # iris = iris.loc[iris['a']>5]
iris

information    a    c                e
samples                               
0            5.1  1.4      Iris-setosa
10           5.4  1.5      Iris-setosa
14           5.8  1.2      Iris-setosa
16           5.4  1.3      Iris-setosa
18           5.7  1.7      Iris-setosa
20           5.4  1.7      Iris-setosa
28           5.2  1.4      Iris-setosa
32           5.2  1.5      Iris-setosa
36           5.5  1.3      Iris-setosa
44           5.1  1.9      Iris-setosa
46           5.1  1.6      Iris-setosa
48           5.3  1.5      Iris-setosa
50           7.0  4.7  Iris-versicolor
52           6.9  4.9  Iris-versicolor
54           6.5  4.6  Iris-versicolor
56           6.3  4.7  Iris-versicolor
58           6.6  4.6  Iris-versicolor
62           6.0  4.0  Iris-versicolor
64           5.6  3.6  Iris-versicolor
66           5.6  4.5  Iris-versicolor
68           6.2  4.5  Iris-versicolor
70           5.9  4.8  Iris-versicolor
72           6.3  4.9  Iris-versicolor
74           6.4  4.3  Iris-versicolor
76           6.8  4.8  Iris-versicolor
78           6.0  4.5  Iris-versicolor
80           5.5  3.8  Iris-versicolor
82           5.8  3.9  Iris-versicolor
84           5.4  4.5  Iris-versicolor
86           6.7  4.7  Iris-versicolor
88           5.6  4.1  Iris-versicolor
90           5.5  4.4  Iris-versicolor
92           5.8  4.0  Iris-versicolor
94           5.6  4.2  Iris-versicolor
96           5.7  4.2  Iris-versicolor
98           5.1  3.0  Iris-versicolor
100          6.3  6.0   Iris-virginica
102          7.1  5.9   Iris-virginica
104          6.5  5.8   Iris-virginica
108          6.7  5.8   Iris-virginica
110          6.5  5.1   Iris-virginica
112          6.8  5.5   Iris-virginica
114          5.8  5.1   Iris-virginica
116          6.5  5.5   Iris-virginica
118          7.7  6.9   Iris-virginica
120          6.9  5.7   Iris-virginica
122          7.7  6.7   Iris-virginica
124          6.7  5.7   Iris-virginica
126          6.2  4.8   Iris-virginica
128          6.4  5.6   Iris-virginica
130          7.4  6.1   Iris-virginica
132          6.4  5.6   Iris-virginica
134          6.1  5.6   Iris-virginica
136          6.3  5.6   Iris-virginica
138          6.0  4.8   Iris-virginica
140          6.7  5.6   Iris-virginica
142          5.8  5.1   Iris-virginica
144          6.7  5.7   Iris-virginica
146          6.3  5.0   Iris-virginica
148          6.2  5.4   Iris-virginica

## 12. reindex를 활용해서 iris의 index를 0부터 149까지로 수정하고, columns는 a와 c만 남겨놓으세요. 결측치는 그 이전 값으로 채우세요.

In [23]:
iris = iris.reindex(index=np.arange(150), columns=['a', 'c'], method='ffill')
iris

information    a    c
samples              
0            5.1  1.4
1            5.1  1.4
2            5.1  1.4
3            5.1  1.4
4            5.1  1.4
...          ...  ...
145          6.7  5.7
146          6.3  5.0
147          6.3  5.0
148          6.2  5.4
149          6.2  5.4

[150 rows x 2 columns]

## 13. apply 메서드를 활용해서, 각 sample 별로 'a' 피처와 'c' 피처의 평균값과, 둘 중 더 큰 값을 각각 'mean'과 'max'라는 이름의 피처로 갖도록 하는 데이터 프레임을 만드세요.

In [25]:
# iris.apply(lambda x: Series([np.mean(x), np.max(x)], index=['mean', 'max']), axis='columns')
mean_max = DataFrame(index=iris.index)
mean_max['mean'] = iris[['a', 'c']].apply(np.mean, axis=1)
mean_max['max'] = iris[['a', 'c']].apply(np.max, axis=1)

mean_max

mean  max
samples           
0        3.25  5.1
1        3.25  5.1
2        3.25  5.1
3        3.25  5.1
4        3.25  5.1
...       ...  ...
145      6.20  6.7
146      5.65  6.3
147      5.65  6.3
148      5.80  6.2
149      5.80  6.2

[150 rows x 2 columns]